### PR TITLE
feature: Support AMDGPU Data Collection

### PR DESCRIPTION
**Pull request by yretenai**
## Description

Adds GPU metrics gathering ~~via amdgpu_top's [libamdgpu_top crate](https://crates.io/crates/libamdgpu_top)~~ on Linux.

image

Some notes: the library queries /proc/{pid}/fdinfo, which can probably be parsed without amdgpu_top's libraries. Intel Arc apparently also uses this fdinfo but I cannot confirm. 

## Testing

_If relevant, please state how this was tested. All changes **must** be tested to work:_

_If this is a code change, please also indicate which platforms were tested:_

- [ ] _Windows_
- [ ] _macOS_
- [x] _Linux_

## Checklist

_If relevant, ensure the following have been met:_

- [x] _Areas your change affects have been linted using rustfmt (`cargo fmt`)_
- [x] _The change has been tested and doesn't appear to cause any unintended breakage_
- [x] _Documentation has been added/updated if needed (`README.md`, help menu, doc pages, etc.)_
- [ ] _The pull request passes the provided CI pipeline_
- [ ] _There are no merge conflicts_
- [ ] _If relevant, new tests were added (don't worry too much about coverage)_

### CODE REVIEW

**Comment by yretenai** [comment]
CI failing due to lacking libdrm/libdrm_amdgpu libraries. What's the best way forward here?

**Comment by ClementTsang** [comment]
Can the libraries be installed?

**Comment by yretenai** [comment]
libdrm should exist in most linux distros via their package managers, i'm atm investigating a dependency-free solution by parsing `/proc/*/fdinfo/*`

**Comment by yretenai** [comment]
Most recent commit parses AMD GPU metrics via procfs (for per-process utilization and video memory usage) and sysfs (for overall AMDGPU memory usage and temperature sensors) and as such doesn't rely on any libraries.

However the code is significantly more complex.

**Review by ClementTsang — COMMENTED** [review]
Left a few comments for now, mostly looks good though.

File: src/data_collection/amd.rs [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+        return None;
+    };
+
+    Some(AMDGPUMemory {
+        total: vram_total,
+        used: vram_used,
+    })
+}
+
+pub fn get_amd_temp(device_path: &Path) -> Option<Vec<AMDGPUTemperature>> {
```
[the remark this answers is not in the document]
Comment (yretenai) [reply src/data_collection/amd.rs:145]: Besides having better names, no. They end up symlinking to the same hwmon endpoint

File: src/data_collection/amd.rs [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+    amdgpu_marketing::AMDGPU_MARKETING_NAME
+        .iter()
+        .find(|(did, rid, _)| (did, rid) == (&device_id, &revision_id))
+        .map(|tuple| tuple.2.to_string())
+}
+
+pub fn get_amd_vram(device_path: &Path) -> Option<AMDGPUMemory> {
+    // get vram memory info from sysfs
+    let vram_total_path = device_path.join("mem_info_vram_total");
+    let vram_used_path = device_path.join("mem_info_vram_used");
```
[the remark this answers is not in the document]
Comment (yretenai) [reply src/data_collection/amd.rs:119]: I do not, but I can ask around.

File: src/data_collection/amd.rs [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+    pub enc_usage: u64,
+    pub dec_usage: u64,
+    pub uvd_usage: u64,
+    pub vcn_usage: u64,
+    pub vpe_usage: u64,
+    pub compute_usage: u64,
+}
+
+// needs previous state for usage calculation
+static PROC_DATA: LazyLock<Mutex<HashMap<PathBuf, HashMap<u32, AMDGPUProc>>>> =
```
[the remark this answers is not in the document]
Comment (yretenai) [reply src/data_collection/amd.rs:49]: It's the only way I could get a mutable reference to PROC_DATA without making the static itself mutable, and thus requiring unsafe code to access.

**Comment by yretenai** [comment]
> Left a few comments for now, mostly looks good though.

I will make the necessary changes tomorrow.

**Comment by yretenai** [comment]
Force pushes were rewording commit messages.

**Comment by yretenai** [comment]
I accidentally reset the signature of the 4th commit from HEAD, which I just fixed by resetting the entire branch, apologies! History should be preserved now.

**Comment by ClementTsang** [comment]
Sorry for the delay, was on vacation - will take a look at this again in a sec.

File: src/data_collection/amd.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ClementTsang) [reply src/data_collection/amd.rs:145]: Hm, guess it's fine to leave it in for now but I might integrate this into that code path in a later PR.

**Review by ClementTsang — APPROVED** [review]
LGTM, thanks for doing this!

**Comment by ClementTsang** [comment]
@all-contributors please add @yretenai for code.

**Pull request merged**